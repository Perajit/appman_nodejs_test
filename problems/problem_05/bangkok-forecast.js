require('isomorphic-fetch');

const { readFileSync } = require('fs');
const API_ROOT = 'http://api.openweathermap.org/data/2.5/forecast/daily';
const bkkId = 1609350;

const apiFilePath = __dirname + '/openweather.key.txt';
const apiKey = readFileSync(apiFilePath, 'utf8');

const request = (url, requestOptions) => {
  return fetch(url, requestOptions)
    .then((res) => res.json())
}

const bangkokForecast = () => {
  return request(`${API_ROOT}?APPID=${apiKey}&units=metric&cnt=7&id=${bkkId}`)
    .then((json) => {
      let { list } = json;

      return list.map((item) => {
        let {
          dt,
          temp: { min, max }
        } = item;
        let date = new Date(dt);

        return {
          date: date.toISOString().substr(0, 10),
          minTemp: min,
          maxTemp: max
        }
      });
    });
};

module.exports = { bangkokForecast };
