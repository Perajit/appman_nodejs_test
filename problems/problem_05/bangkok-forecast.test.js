const { expect } = require('chai');
const { stub } = require('sinon');
const src = require('./bangkok-forecast');
const mock = require('./bangkok-forecast.test.mock');
const expected = require('./bangkok-forecast.test.expected');

const mockRequest = (mock) => {
  return Promise.resolve({
    json: () => mock
  });
}

describe('bangkok-forecase', () => {
  describe('bangkokForecast', () => {
    beforeEach(() => {
      this.fetch = stub(global, 'fetch');
      this.fetch.returns(mockRequest(mock));

      this.convertData = stub(src, 'convertData');
      this.convertData.returns(expected);
    });

    afterEach(() => {
      global.fetch.restore();
    });

    it('should return converted data from api', (done) => {
      src.bangkokForecast()
        .then((actualResult) => {
          expect(actualResult).to.deep.equal(expected);
          done();
        });
    });
  });

  describe('convertData', () => {
    it('should convert data correctly', () => {
      let actualResult = src.convertData(mock);
      expect(actualResult).to.deep.equal(expected);
    });
  });
});

