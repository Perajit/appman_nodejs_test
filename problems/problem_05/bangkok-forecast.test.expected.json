[
  {
    "date": "2018-04-21",
    "minTemp": 27.34,
    "maxTemp": 29
  },
  {
    "date": "2018-04-22",
    "minTemp": 27.61,
    "maxTemp": 35.38
  },
  {
    "date": "2018-04-23",
    "minTemp": 27.94,
    "maxTemp": 36.35
  },
  {
    "date": "2018-04-24",
    "minTemp": 28.19,
    "maxTemp": 37.17
  },
  {
    "date": "2018-04-25",
    "minTemp": 27.88,
    "maxTemp": 36.23
  },
  {
    "date": "2018-04-26",
    "minTemp": 26.72,
    "maxTemp": 36.2
  },
  {
    "date": "2018-04-27",
    "minTemp": 25.71,
    "maxTemp": 34.49
  }
]
